crypto/sha/asm/keccak1600-armv4.pl

```diff
@@ -196,409 +204,419 @@
 	eor	@C[2],@C[2],@C[7],ror#32-1	@ C[1] = ROL64(C[3], 1) ^ C[1];
 	str	@C[1],[sp,#$D[0]+4]
 	eor	@C[3],@C[3],@C[6]
+	 ldr	@C[7],[sp,#$A[3][3]]
 	str	@C[2],[sp,#$D[2]]		@ D[2] = C[1]
 	eor	@C[4],@C[4],@C[9],ror#32-1	@ C[2] = ROL64(C[4], 1) ^ C[2];
+	 ldr	@C[6],[sp,#$A[3][3]+4]
 	str	@C[3],[sp,#$D[2]+4]
 	eor	@C[5],@C[5],@C[8]
-	 ldr	@C[8],[sp,#$A[3][0]]
-	 ldr	@C[9],[sp,#$A[3][0]+4]
-	str	@C[4],[sp,#$D[3]]		@ D[3] = C[2]
-	str	@C[5],[sp,#$D[3]+4]
-
-	ldr	@C[6],[sp,#$A[0][1]]
-	eor	@C[8],@C[8],@C[0]
-	ldr	@C[7],[sp,#$A[0][1]+4]
-	eor	@C[9],@C[9],@C[1]
-	str	@C[8],[sp,#$T[0][0]]		@ T[0][0] = A[3][0] ^ C[0]; /* borrow T[0][0] */
-	ldr	@C[8],[sp,#$A[0][2]]
-	str	@C[9],[sp,#$T[0][0]+4]
-	ldr	@C[9],[sp,#$A[0][2]+4]
-	eor	@C[6],@C[6],@E[0]
-	eor	@C[7],@C[7],@E[1]
-	str	@C[6],[sp,#$T[0][1]]		@ T[0][1] = A[0][1] ^ E[0]; /* D[1] */
-	ldr	@C[6],[sp,#$A[0][3]]
-	str	@C[7],[sp,#$T[0][1]+4]
-	ldr	@C[7],[sp,#$A[0][3]+4]
-	eor	@C[8],@C[8],@C[2]
-	eor	@C[9],@C[9],@C[3]
-	str	@C[8],[sp,#$T[0][2]]		@ T[0][2] = A[0][2] ^ C[1]; /* D[2] */
-	ldr	@C[8],[sp,#$A[0][4]]
-	str	@C[9],[sp,#$T[0][2]+4]
-	ldr	@C[9],[sp,#$A[0][4]+4]
-	eor	@C[6],@C[6],@C[4]
-	eor	@C[7],@C[7],@C[5]
-	str	@C[6],[sp,#$T[0][3]]		@ T[0][3] = A[0][3] ^ C[2]; /* D[3] */
-	eor	@C[8],@C[8],@E[2]
-	str	@C[7],[sp,#$T[0][3]+4]
-	eor	@C[9],@C[9],@E[3]
-	 ldr	@C[6],[sp,#$A[3][3]]
-	 ldr	@C[7],[sp,#$A[3][3]+4]
-	str	@C[8],[sp,#$T[0][4]]		@ T[0][4] = A[0][4] ^ E[1]; /* D[4] */
-	str	@C[9],[sp,#$T[0][4]+4]
 
 	ldr	@C[8],[sp,#$A[4][4]]
-	eor	@C[4],@C[4],@C[6]
 	ldr	@C[9],[sp,#$A[4][4]+4]
-	eor	@C[5],@C[5],@C[7]
-	ror	@C[7],@C[4],#32-10		@ C[3] = ROL64(A[3][3] ^ C[2], rhotates[3][3]);   /* D[3] */
+	 str	@C[4],[sp,#$D[3]]		@ D[3] = C[2]
+	eor	@C[7],@C[7],@C[4]
+	 str	@C[5],[sp,#$D[3]+4]
+	eor	@C[6],@C[6],@C[5]
 	ldr	@C[4],[sp,#$A[0][0]]
-	ror	@C[6],@C[5],#32-11
-	ldr	@C[5],[sp,#$A[0][0]+4]
+	@ ror	@C[7],@C[7],#32-10		@ C[3] = ROL64(A[3][3] ^ C[2], rhotates[3][3]);   /* D[3] */
```

Comment:
Commit message mentions "merge of rotations with logical instructions". What does it mean in practice? If you have `ror x,y,#z` followed by say `or a,b,x`, [on ARM] you can replace them with single `or a,b,y,ror#z`. That's what happens here, and commented ror instruction is kind of a "forward reference" or a hint on what to look for, e.g. in case trace `@C[7]`. You can wonder if it's possible to merge *all* of them and not just "almost 1/2". Challenge there is that you have operations like `or a,b>>>x,c>>>y`, i.e. both input operand that need to be rotated. It's possible to solve this by `or a,b,c>>>(x-z)` and then rotate `a` accordingly in next operation that takes `a` as input. But trouble is that next operation on `a` can be in next round, which means that you'd have to store rotated value in memory. It's totally possible to do, but it increases complexity (since it's harder to trace rotate factors between rounds). So I decided to *not* do it [yet?], which leaves "almost 1/2" merged rotates, those that don't have to be stored in memory between rounds.

As for "[yet?]". One has to keep in mind that majority of contemporary processors, including micro-controllers, have NEON engine and NEON code path is [and will always be] significantly faster. Or in other words the code path in question essentially targets legacy processors, and then question is if it actually makes a lot of sense to invest that much effort into code that will be hardly used. On related note one can add that there is another option that might give *larger* boost, but it's quite "narrow". If one claims that ARMv6 with Thumb-2, or ARMv7 micro-controllers without NEON engine, are extremely important, the it's possible to switch to double-word loads and stores and reduce amount of instructions by almost 20%, which would result in >20% performance improvement. For reference, folding remaining rotates would give ~5%. [Well, one can use double-word loads and stores even without Thumb-2, but this is post-ARMv4 instruction and without Thumb-2 one can't replace as many of them.]